cmd/update-rules/main.go

```diff
@@ -0,0 +1,162 @@
+
+  # Update rules and export to /tmp/updated_rules.yaml for release-1.22 branch with go version version 1.16.3
+  update-rules -branch release-1.22 -go 1.16.3 -o /ymp/updated_rules.yaml -rule /go/src/k8s.io/kubernetes/staging/publishing/rules.yaml`
+
+	flag.Usage = func() {
+		fmt.Fprintf(os.Stdout, "\n  Usage:  update-rules %s %s [%s | %s]", branchExample, ruleExample, "-go", "-o")
+		fmt.Fprintf(os.Stdout, "\n  %s\n\n", examples)
+		flag.PrintDefaults()
+	}
+
+	branch := flag.String("branch", "", branchUsage)
+	rulesFile := flag.String("rule", "", ruleUsage)
+	goVersion := flag.String("go", "", goVerUsage)
+	out := flag.String("o", "", outUsage)
+	// TODO: add dryrun option
+
+	flag.Parse()
+
+	if *branch == "" {
+		glog.Errorf("branch flag requires a non-empty value, for example: %s", branchExample)
+		os.Exit(2)
+	}
+	// TODO: add branch validation
+
+	if *rulesFile == "" {
+		glog.Errorf("rule flag requires a non-empty value, for example: %s", ruleExample)
+		os.Exit(2)
+	}
+
+	var outFile string
+	if *out != "" {
+		outFile = *out
+	} else {
+		outFile = *rulesFile
+	}
+
+	rules, err := config.LoadRules(*rulesFile)
+	if err != nil {
+		glog.Fatalf("can not load rules file %q: %v", *rulesFile, err)
+	}
+
+	if err := config.Validate(rules); err != nil {
+		glog.Fatalf("invalid rule file %q: %v", *rulesFile, err)
+	}
+
+	// use the golang version specified via input or default in input rule file
+	if *goVersion == "" {
+		*goVersion = *rules.DefaultGoVersion
+	}
+
+	for j, r := range rules.Rules {
+		var mainBranchRuleFound bool
+		var newBranchRule config.BranchRule
+		// find the mainBranch rule
+		for _, br := range r.Branches {
+			if br.Name == mainBranch {
+				newBranchRule = config.BranchRule(br)
+				mainBranchRuleFound = true
+				break
+			}
+		}
+		// if mainBranch rule not found for target repo, error
+		if !mainBranchRuleFound {
+			glog.Fatalf("%s branch rule not found for repo %s", mainBranch, r.DestinationRepository)
+		}
+
+		// copy the mainBranch rule and update the branch name and go version
+		newBranchRule.Name = *branch
+		newBranchRule.Source.Branch = *branch
+		// pin the go version per branch rule update
+		if newBranchRule.GoVersion == "" {
+			newBranchRule.GoVersion = *goVersion
+		}
+
+		var branchRuleExists bool
+		// if the target branch rule already exists, update it
+		for i, br := range r.Branches {
+			if br.Name == *branch {
+				glog.Infof("found branch %s rule for repo %s, updating it", *branch, r.DestinationRepository)
+				r.Branches[i] = newBranchRule
+				branchRuleExists = true
+				break
+			}
+		}
+		// new rule, append to destination's branches
+		if !branchRuleExists {
+			r.Branches = append(r.Branches, newBranchRule)
+		}
+
+		// update the rule for destination repo
+		rules.Rules[j] = r
+	}
+
+	// validate rules after update
+	if err := config.Validate(rules); err != nil {
+		glog.Fatalf("update failed, found invalid rules after update: %v", err)
+	}
+
+	data, err := yaml.Marshal(rules)
```

Comment:
This removes comments...maybe consider using go-yaml/yaml to preserve comments (https://github.com/go-yaml/yaml/issues/132)

Reply:
The comments are removed while the yaml content are deserialized into native ds. So at this point we don't have comments at all. To preserve comments, we'd need to load them into `yaml.Node{}` and navigate using it (we may also need to update the way rules loading and validations are done in `cmd/publishing-bot/config/rules.go`). It's available in yaml.v3, currently we are using yaml.v2 here.